blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +385,219 @@ where
+        })?;
+
+        let tipset = match first_subscriber {
+            HeadChange::Current(tipset) => tipset,
+            _ => {
+                return Err(Error::Other(format!(
+                    "expected current head on SHC stream (got {:?})",
+                    first_subscriber
+                )))
+            }
+        };
+        let maybe_message_reciept = self.tipset_executed_message(&tipset, cid, &*message)?;
+        if let Some(r) = maybe_message_reciept {
+            return Ok(Some((tipset.clone(), r)));
+        }
+
+        let (back_tipset, _back_receipt) = self
+            .search_back_for_message(&tipset, &*message)?
+            .ok_or_else(|| {
+                Error::Other("State manager not subscribed to back search wait".to_string())
+            })?;
+
+        let mut candidate_tipset: Option<Arc<Tipset>> = None;
+        let mut candidate_receipt: Option<MessageReceipt> = None;
+        let height_of_head = tipset.epoch();
+        let mut reverts: HashMap<TipsetKeys, bool> = HashMap::new();
+
+        while let Some(subscriber) = self
+            .subscriber
+            .clone()
+            .ok_or_else(|| {
+                Error::Other("State Manager not subscribed to tipset head changes".to_string())
+            })?
+            .next()
+            .await
+        {
+            match subscriber {
+                HeadChange::Revert(_tipset) => {
+                    if candidate_tipset.is_some() {
+                        candidate_tipset = None;
+                        candidate_receipt = None;
+                    }
+                }
+                HeadChange::Apply(tipset) => {
+                    if candidate_tipset
+                        .as_ref()
+                        .map(|s| s.epoch() >= s.epoch() + tipset.epoch())
+                        .unwrap_or_default()
+                    {
+                        let ts = candidate_tipset
+                            .ok_or_else(|| Error::Other("Candidate Tipset not".to_string()))?;
+
+                        let rs = candidate_receipt
+                            .ok_or_else(|| Error::Other("Candidate Receipt not set".to_string()))?;
+
+                        return Ok(Some((ts, rs)));
+                    }
+
+                    reverts.insert(tipset.key().to_owned(), true);
+
+                    let maybe_receipt = self.tipset_executed_message(&tipset, cid, &*message)?;
+                    if let Some(receipt) = maybe_receipt {
+                        if confidence == 0 {
+                            return Ok(Some((tipset, receipt)));
+                        }
+                        candidate_tipset = Some(tipset);
+                        candidate_receipt = Some(receipt)
+                    }
+                }
+                _ => (),
+            }
+        }
+
+        while back_search_wait.next().await.is_some() {
+            if !reverts.get(back_tipset.key()).unwrap_or(&false)
```

Comment:
This gets called for all `()` put through this pubsub channel, but this `back_tipset` never changes, would this not just loop over and call the same code on the tipset when the `back_tipset` was initiated?

Reply:
I thought since it is in a loop structure the same thing would be happening at the lotus side? If multiple backsearchWait events are being called the same code will potentially be called as well?

Reply:
Well, they are just calling that function asyncronously and only handling it once (and not recurring, they kinda use that channel as a oneshot. Seems like if you have this subscriber you could potentially be receiving events from another function call, unrelated to the logic of this. 

I don't see why these events would need to be published in a pubsub way, seems pretty local (maybe clarification around why this needs to be emitted like this would be great, as I'm confused)

Reply:
Definitely! Wanted to use a oneshot but it looks like subheadchanges is the one that is sending these events externally? That is where the tsub starts to match in the case? I was also confused as to why they need to be externally called but following lotus seems like that might be the way they want it? I could also be very confused it is not as easy to parse haha

https://github.com/filecoin-project/lotus/blob/c4de2373fd13c909876cf459e59ead7749b9939d/chain/stmgr/stmgr.go#L494

Reply:
Yes, I see they are using that and that is matched above, but this back search is completely distinct and you are using the same subscriber channel, this will just keep iterating over and never close, so this function will never end. Am I missing something here?

Reply:
this is my lack of go showing, I just recently knew that go func() is supposed to be an anoymous routine. Taking a look at this and fixing this right now, was trying to visualize as to what point the back_search_wait gets sent to and this makes more sense now